### PR TITLE
Add duplicate method to collection, sample and data resource

### PR DESCRIPTION
**Pull request by hadalin**
This seems to be working, but it needs some polishing:

- Better output when when a resource is duplicated. Currently the whole duplicated object (dict) is returned
- Tests
- Changelog entry
- SDK Reference?

### CODE REVIEW

File: resdk/resources/collection.py [hunk 1]
```diff
@@ -264,3 +264,11 @@ def remove_samples(self, *samples):
...
 
         self.samples.clear_cache()
+
+    def duplicate(self):
+        """Duplicate (make copy of) ``collection`` object."""
+        if self.id is None:
+            raise ValueError('Instance must be saved before making a duplicate.')
+
+        duplicated = self.resolwe.api.collection().duplicate.post({'ids': [self.id]})
+        return duplicated[0]
```
Comment (dblenkus) [inline resdk/resources/collection.py]: You need to return a `Collection` object, something like `self.__class__(resolwe=self.resolwe, **duplicated[0])`

**Comment by dblenkus** [comment]
And don't forget to add a change log.

**Comment by dblenkus** [comment]
This looks good!

Sample endpoint also support [`inherit_collection`](https://github.com/genialis/resolwe/blob/c401f8363e10307cb2c507a1c8ec733cc1599e1e/resolwe/flow/views/entity.py#L103) parameter. Can you also support that? And don't forget to handle errors that can be raised.

File: resdk/tests/functional/duplication/e2e_duplication.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# pylint: disable=missing-docstring
+from resdk.tests.functional.base import BaseResdkFunctionalTest
+
+
+class TestDuplication(BaseResdkFunctionalTest):
+
+    def test_collection_duplication(self):
```
Comment (hadalin) [inline resdk/tests/functional/duplication/e2e_duplication.py]: @dblenkus travis-ci doesn't seem to [run these tests](https://travis-ci.org/genialis/resolwe-bio-py/jobs/565663036#L406-L413). Is this expected?

File: resdk/resources/sample.py [hunk 3]
```diff
@@ -313,3 +313,22 @@ def is_background(self):
             self._is_background = len(background_relations) > 0  # pylint: disable=len-as-condition
 
         return self._is_background
+
+    def duplicate(self, inherit_collections=False):
+        """Duplicate (make copy of) ``sample`` object.
+
+        :param inherit_collections: If ``True`` then duplicated
```
Comment (hadalin) [inline resdk/resources/sample.py]: Added `inherit_collections` parameter.

**Comment by hadalin** [comment]
>And don't forget to handle errors that can be raised.

Exceptions seem to be handled with `handle_http_exception`. Should we leave exception handling to the end user, instead of doing it for them?

File: resdk/resources/sample.py (hunk 3 above)
Comment (dblenkus) [reply resdk/resources/sample.py]: This was renamed to `inherit_collection` in latest resolwe.

**Comment by dblenkus** [comment]
It's fine if an error is raised, we just want to make sure it includes relevant informations about the underlying error.

If api returns meaningful response, you should use it for a user-friendly exception. Otherwise we will handle this when we will improve error handling in general, but that will require quite some refactoring on the backend.

**Comment by hadalin** [comment]
What do you have in mind? Creating `DuplicationError(Exception)` and packing the original exception in it?

File: resdk/tests/functional/duplication/e2e_duplication.py (hunk 2 above)
Comment (hadalin) [reply resdk/tests/functional/duplication/e2e_duplication.py]: I think I found the problem. There no `__init__.py` file in the directory.

Comment (hadalin) [reply resdk/tests/functional/duplication/e2e_duplication.py]: Though is seems that not one functional test is actually tested.

Comment (hadalin) [reply resdk/tests/functional/duplication/e2e_duplication.py]: Oh, never mind, I think Jenkins runs them.

Comment (JureZmrzlikar) [reply resdk/tests/functional/duplication/e2e_duplication.py]: Yes, e2e tests are run by Jenkins.

File: resdk/tests/functional/duplication/e2e_duplication.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        collection = self.res.collection.create(name='Test collection')
+        duplicate = collection.duplicate()
+        self.assertEqual(duplicate.name, 'Copy of Test collection')
+
+    def test_sample_duplication(self):
+        sample = self.res.sample.create(name='Test sample')
+        duplicate = sample.duplicate()
+        self.assertEqual(duplicate.name, 'Copy of Test sample')
+
+    def test_sample_duplication_perms(self):
```
Comment (JureZmrzlikar) [inline resdk/tests/functional/duplication/e2e_duplication.py]: IMO, this test does not belong here. In resdk, we should only verify that Resolwe endpoints work with a minimal example (e.g. creating a duplicated data/sample/collection). All edge cases, permissions, etc... should be tested in Resolwe.

File: resdk/tests/functional/duplication/e2e_duplication.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        duplicate = sample.duplicate(inherit_collection=True)
+        self.assertEqual(len(collection.samples), 1)
+        self.assertEqual(collection.samples[0].name, 'Test sample')
+        self.assertEqual(duplicate.name, f'Copy of {sample.name}')
+
+    def test_data_duplication(self):
+        data = self.res.run(slug='test-sleep-progress', input={'t': 1})
+        # Let's not wait for processing to finish and
+        # check the expected exception to be raised.
+        with self.assertRaisesRegex(ResolweServerError, 'done or error status to be duplicated'):
```
Comment (JureZmrzlikar) [inline resdk/tests/functional/duplication/e2e_duplication.py]: Why do you expect error during duplication?

Comment (hadalin) [reply resdk/tests/functional/duplication/e2e_duplication.py]: Read the comment above :)

File: resdk/tests/functional/duplication/e2e_duplication.py (hunk 4 above)
Comment (hadalin) [reply resdk/tests/functional/duplication/e2e_duplication.py]: This test is there to check what happens if permission error is raised. @dblenkus suggested adding a `PermissionError(ResolweServerError)` to `exceptions.py ` and raising that if resolwe returns permission error.

File: resdk/tests/functional/duplication/e2e_duplication.py (hunk 5 above)
Comment (hadalin) [reply resdk/tests/functional/duplication/e2e_duplication.py]: Duplication is allowed only for data objects that have `OK` or `ER` status. In other cases exception is raised and since we only care if endpoint works as expected, we check that.

File: resdk/tests/functional/duplication/e2e_duplication.py (hunk 4 above)
Comment (JureZmrzlikar) [reply resdk/tests/functional/duplication/e2e_duplication.py]: OK, this sounds fine.

File: resdk/tests/functional/duplication/e2e_duplication.py (hunk 5 above)
Comment (JureZmrzlikar) [reply resdk/tests/functional/duplication/e2e_duplication.py]: OK, makes sense.

Comment (JureZmrzlikar) [reply resdk/tests/functional/duplication/e2e_duplication.py]: Maybe add an explicit comment that error is expected because data will not have an OK/ER status. I am familiar with Resolwe duplication mechanism, but still haven't made the connection.

File: resdk/tests/functional/duplication/e2e_duplication.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        sample = self.res.sample.create(name='Test sample')
+        duplicate = sample.duplicate()
+        self.assertEqual(duplicate.name, 'Copy of Test sample')
+
+    def test_sample_duplication_perms(self):
+        collection = self.user_res.collection.create()
+        sample = self.user_res.sample.create(name='Test sample')
+        collection.add_samples(sample)
+
+        collection.permissions.remove_user(self.user_res.auth.username, 'add')
```
Comment (dblenkus) [inline resdk/tests/functional/duplication/e2e_duplication.py]: Use `USER_USERNAME` (form `resdk.tests.functional.base`) instead of `self.user_res.auth.username`

Comment (dblenkus) [reply resdk/tests/functional/duplication/e2e_duplication.py]: Sidenote for @JureZmrzlikar: we should add constants with permission values, e.g. `view`, `edit`, `add`,...

Comment (dblenkus) [reply resdk/tests/functional/duplication/e2e_duplication.py]: @hadalin it would be cleaner approach to create objects with `self.res` (admin user) and then grant permissions to `USER_USERNAME`.

File: resdk/tests/functional/duplication/e2e_duplication.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        # TODO is it ok that no exception is raised when duplicating a sample
+        # with 'inherit_collection' set to True while the user has no 'add' permission
+        # on the collection?
+        # What happens is that the sample is duplicated just fine, but is also NOT
+        # added to the collection (without an exception being raised).
```
Comment (dblenkus) [inline resdk/tests/functional/duplication/e2e_duplication.py]: This should be fixed in Resolwe. @hadalin can you do this?

File: resdk/tests/functional/duplication/e2e_duplication.py (hunk 6 above)
Comment (JureZmrzlikar) [reply resdk/tests/functional/duplication/e2e_duplication.py]: Regarding constants for view, edit, add, ... I'd rather spend effort polishing this at later time -when permissions changes take place.

Comment (dblenkus) [reply resdk/tests/functional/duplication/e2e_duplication.py]: >  at later time

That's why it is a sidenote :)

Comment (hadalin) [reply resdk/tests/functional/duplication/e2e_duplication.py]: >it would be cleaner approach to create objects with self.res

Isn't it required to create a sample like this `sample = self.user_res.sample.create` so that `sample.duplicate()` makes a request as `'user'`?

Also isn't it more obvious what we're trying to test if we _remove_ the permission? If not, then please provide a code snippet, so I can see what you mean.

File: resdk/tests/functional/duplication/e2e_duplication.py (hunk 7 above)
Comment (hadalin) [reply resdk/tests/functional/duplication/e2e_duplication.py]: I could, but it's not a priority :)

I'd merge as it is and deal with this later as this is a separate issue.

File: resdk/tests/functional/duplication/e2e_duplication.py (hunk 5 above)
Comment (hadalin) [reply resdk/tests/functional/duplication/e2e_duplication.py]: Added an extra comment below.

File: resdk/tests/functional/duplication/e2e_duplication.py (hunk 7 above)
Comment (dblenkus) [reply resdk/tests/functional/duplication/e2e_duplication.py]: You can, but it doesn't make sense to add a test that works against expectations :)

Comment (hadalin) [reply resdk/tests/functional/duplication/e2e_duplication.py]: Well, these _are_ the current expectations :)

Either we

- [ ] keep the test which will fail once the issue is fixed in Resolwe, but we get to keep the test and not forget about it
- [ ] remove the test

File: resdk/resources/sample.py [hunk 8]
```diff
@@ -225,3 +225,22 @@ def is_background(self):
             self._is_background = len(background_relations) > 0  # pylint: disable=len-as-condition
 
         return self._is_background
+
+    def duplicate(self, inherit_collection=False):
+        """Duplicate (make copy of) ``sample`` object.
+
+        :param inherit_collection: If ``True`` then duplicated
+            samples will be added to collections the original sample
```
Comment (JureZmrzlikar) [inline resdk/resources/sample.py]: Since samples and collection now have m2o instead of m2m relation, this could be simplified. Something like: "If `True` then duplicated samples (and their data) will be added to collection of the original sample."

File: resdk/resources/collection.py [hunk 9]
```diff
@@ -185,3 +185,14 @@ def relations(self):
             self._relations = self.resolwe.relation.filter(collection=self.id)
 
         return self._relations
+
+    def duplicate(self):
+        """Duplicate (make copy of) ``collection`` object.
+
+        :return: Duplicated collection
+        """
+        if self.id is None:
```
Comment (JureZmrzlikar) [inline resdk/resources/collection.py]: The concept of checking the existence of `id` was so common that we made a [assert_object_exists](https://github.com/genialis/resolwe-bio-py/blob/master/resdk/utils/decorators.py#L24) decorator for it. Please, use it here and in all other `duplicate()` methods.

**Comment by hadalin** [comment]
@JureZmrzlikar I haven't followed what changes were made in Resolwe. E2E tests seems to be failing because of `Bad Request: /api/data/duplicate`. Would you mind giving pointers what needs to be fixed?
